File: Management/Services/Canvas/CanvasAssignmentGroupService.cs
using CanvasModel.Assignments;
using LocalModels;
using RestSharp;

namespace Management.Services.Canvas;


public interface ICanvasAssignmentGroupService
{
  Task<IEnumerable<CanvasAssignmentGroup>> GetAll(ulong courseId);
  Task<LocalAssignmentGroup> Create(ulong canvasCourseId, LocalAssignmentGroup localAssignmentGroup);
  Task Update(ulong canvasCourseId, LocalAssignmentGroup localAssignmentGroup);

}
public class CanvasAssignmentGroupService: ICanvasAssignmentGroupService
{
  private readonly IWebRequestor webRequestor;
  private readonly CanvasServiceUtils utils;
  private readonly MyLogger<CanvasAssignmentGroupService> logger;

  public CanvasAssignmentGroupService(
    IWebRequestor webRequestor,
    CanvasServiceUtils utils,
    MyLogger<CanvasAssignmentGroupService> logger
  )
  {
    this.webRequestor = webRequestor;
    this.utils = utils;
    this.logger = logger;
  }
  public async Task<IEnumerable<CanvasAssignmentGroup>> GetAll(ulong courseId)
  {
    var url = $"courses/{courseId}/assignment_groups";
    var request = new RestRequest(url);
    var assignmentResponse = await utils.PaginatedRequest<IEnumerable<CanvasAssignmentGroup>>(request);
    return assignmentResponse.SelectMany(
      assignments => assignments
    );
  }

  public async Task<LocalAssignmentGroup> Create(
    ulong canvasCourseId,
    LocalAssignmentGroup localAssignmentGroup
  )
  {
    logger.Log($"creating assignment group: {localAssignmentGroup.Name}");
    var url = $"courses/{canvasCourseId}/assignment_groups";
    var request = new RestRequest(url);
    var body = new
    {
      name = localAssignmentGroup.Name,
      group_weight = localAssignmentGroup.Weight,
    };
    request.AddBody(body);

    var (canvasAssignmentGroup, response) = await webRequestor.PostAsync<CanvasAssignmentGroup>(request);
    if (canvasAssignmentGroup == null)
      throw new Exception("created canvas assignment group was null");

    return localAssignmentGroup with
    {
      CanvasId = canvasAssignmentGroup.Id
    };
  }
  public async Task Update(
    ulong canvasCourseId,
    LocalAssignmentGroup localAssignmentGroup
  )
  {
    logger.Log($"updating assignment group: {localAssignmentGroup.Name}");
    if (localAssignmentGroup.CanvasId == null)
      throw new Exception("cannot update assignment group if canvas id is null");
    var url = $"courses/{canvasCourseId}/assignment_groups/{localAssignmentGroup.CanvasId}";
    var request = new RestRequest(url);
    var body = new
    {
      name = localAssignmentGroup.Name,
      group_weight = localAssignmentGroup.Weight,
    };
    request.AddBody(body);

    await webRequestor.PutAsync<CanvasAssignmentGroup>(request);
  }
}

File: src/features/canvas/services/canvasAssignmentGroupService.ts
import { canvasApi, paginatedRequest } from "./canvasServiceUtils";
import { CanvasAssignmentGroup } from "@/features/canvas/models/assignments/canvasAssignmentGroup";
import { LocalAssignmentGroup } from "@/features/local/assignments/models/localAssignmentGroup";
import { rateLimitAwareDelete, rateLimitAwarePost } from "./canvasWebRequestUtils";
import { axiosClient } from "@/services/axiosUtils";

export const canvasAssignmentGroupService = {
  async getAll(courseId: number): Promise<CanvasAssignmentGroup[]> {
    console.log("Requesting assignment groups");
    const url = `${canvasApi}/courses/${courseId}/assignment_groups`;
    const assignmentGroups = await paginatedRequest<CanvasAssignmentGroup[]>({
      url,
    });
    return assignmentGroups.flatMap((groupList) => groupList);
  },

  async create(
    canvasCourseId: number,
    localAssignmentGroup: LocalAssignmentGroup
  ): Promise<LocalAssignmentGroup> {
    console.log(`Creating assignment group: ${localAssignmentGroup.name}`);
    const url = `${canvasApi}/courses/${canvasCourseId}/assignment_groups`;
    const body = {
      name: localAssignmentGroup.name,
      group_weight: localAssignmentGroup.weight,
    };

    const { data: canvasAssignmentGroup } =
      await rateLimitAwarePost<CanvasAssignmentGroup>(url, body);

    return {
      ...localAssignmentGroup,
      canvasId: canvasAssignmentGroup.id,
    };
  },

  async update(
    canvasCourseId: number,
    localAssignmentGroup: LocalAssignmentGroup
  ): Promise<void> {
    console.log(
      `Updating assignment group: ${localAssignmentGroup.name}, ${localAssignmentGroup.canvasId}`
    );
    if (!localAssignmentGroup.canvasId) {
      throw new Error("Cannot update assignment group if canvas ID is null");
    }
    const url = `${canvasApi}/courses/${canvasCourseId}/assignment_groups/${localAssignmentGroup.canvasId}`;
    const body = {
      name: localAssignmentGroup.name,
      group_weight: localAssignmentGroup.weight,
    };

    await axiosClient.put(url, body);
  },

  async delete(
    canvasCourseId: number,
    canvasAssignmentGroupId: number,
    assignmentGroupName: string
  ): Promise<void> {
    console.log(`Deleting assignment group: ${assignmentGroupName}`);
    const url = `${canvasApi}/courses/${canvasCourseId}/assignment_groups/${canvasAssignmentGroupId}`;
    await rateLimitAwareDelete(url);
  },
};
